internal/log/hook.go

```
@@ -1,35 +1,162 @@
+	if !(h.EncodeAsJSON || formatTime) {
+		return
+	}
+
+	// todo: replace these with constraints.Integer, constraints.Float, etc in go1.18
+	for k, v := range d {
+		if !h.EncodeError {
+			if _, ok := v.(error); k == logrus.ErrorKey || ok {
+				continue
+			}
+		}
+
+		if t, ok := v.(time.Time); formatTime && ok {
+			d[k] = t.Format(h.TimeFormat)
+			continue
+		}
+
+		if !h.EncodeAsJSON {
+			continue
+		}
+
+		switch vv := v.(type) {
+		// built in types
+		case bool, string, error, uintptr,
+			int8, int16, int32, int64, int,
+			uint8, uint32, uint64, uint,
+			float32, float64:
+			continue
+
+		case time.Duration:
+			if h.DurationFormat != nil {
+				if i := h.DurationFormat(vv); i != nil {
+					d[k] = i
+				}
+			}
+			continue
+
+		// Rather than setting d[k] = vv.String(), JSON encode []byte value, since it
+		// may be a binary payload and not representable as a string.
+		// `case bytes.Buffer,*bytes.Buffer:` resolves `vv` to `interface{}`,
+		// so cannot use `vv.Bytes`.
+		// Could move to below the `reflect.Indirect()` call below, but
+		// that would require additional typematching and dereferencing.
+		// Easier to keep these duplicate branches here.
+		case bytes.Buffer:
+			v = vv.Bytes()
+		case *bytes.Buffer:
+			v = vv.Bytes()
+		}
+
+		// dereference any pointers
+		rv := reflect.Indirect(reflect.ValueOf(v))
+		// check if `v` is a null pointer
+		if !rv.IsValid() {
+			d[k] = nullString
+			continue
+		}
+
+		switch rv.Kind() {
+		case reflect.Map, reflect.Struct, reflect.Array, reflect.Slice:
+		default:
+			continue
+		}
+
+		b, err := encode(v)
+		if err != nil {
+			// Errors are written to stderr (ie, to `panic.log`) and stops the remaining
+			// hooks (ie, exporting to ETW) from firing. So add encoding errors to
+			// the entry data to be written out, but keep on processing.
+			d[k+"-"+logrus.ErrorKey] = err.Error()
+		}
+
+		// if  `err != nil`, then `b == nil` and this will be the empty string
+		d[k] = string(b)
```

If encoding failed, perhaps better to keep `d[k]` as its original value, rather than overwriting it with an empty string?

---

as in, place the original struct in the key instead? I can do that